cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+  CUDAContextManager context_manager;
+};
+
+using data_t = struct data_s;
+
+void set_entry_ts(data_t *data, hpt_t hpt, int64_t ts) {
+  data->entry_ts[hpt] = ts;
+}
+
+int64_t get_entry_ts(data_t *data, hpt_t hpt) {
+  return data->entry_ts.at(hpt);
+}
+
+std::string strip_event_class_name(const char *str) {
+  std::string temp(str);
+  std::smatch match;
+  std::regex_search(temp, match, std::regex(":(.*?)_?(?:entry|exit)?$"));
+
+  // The entire match is hold in the first item, sub_expressions 
+  // (parentheses delimited groups) are stored after.
+  assert( match.size() > 1 && "Event_class_name not matching regex.");
+
+  return match[1].str();
+}
+
+static void send_host_message(void *btx_handle, void *usr_data, int64_t ts,
+                              const char *event_class_name, const char *hostname, int64_t vpid,
+                              uint64_t vtid, bool err) {
+  std::string event_class_name_striped = strip_event_class_name(event_class_name);
+  const int64_t _start = get_entry_ts(static_cast<data_t *>(usr_data), {hostname, vpid, vtid});
+
+  btx_push_message_lttng_host(btx_handle, hostname, vpid, vtid, _start, BACKEND_CUDA,
+                              event_class_name_striped.c_str(), (ts - _start), err);
+}
+
+static void send_traffic_message(void *btx_handle, void *usr_data, int64_t ts,
+                                 const char *event_class_name, const char *hostname,
+                                 int64_t vpid, uint64_t vtid, bool err) {
+  std::string event_class_name_stripped = strip_event_class_name(event_class_name);
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key = {hostname, vpid, vtid};
+  const int64_t _start = get_entry_ts(state, key);
+
+  if (!err) {
+    auto size = state->entry_traffic_size.at(key);
+    btx_push_message_lttng_traffic(btx_handle, hostname, vpid, vtid, _start, BACKEND_CUDA,
+                                   event_class_name_stripped.c_str(), size);
+  }
+}
+
+
+void btx_initialize_component(void **usr_data) { *usr_data = new data_t; }
+
+void btx_finalize_component(void *usr_data) {
+  delete static_cast<data_t *>(usr_data);
+}
+
+void entries_callback(void *btx_handle, void *usr_data, int64_t ts,
+                      const char *event_class_name, const char *hostname, int64_t vpid,
+                             uint64_t vtid) {
+  set_entry_ts(((data_t *)usr_data), {hostname, vpid, vtid}, ts);
+}
+
+void exits_callback_cudaError_absent(void *btx_handle, void *usr_data, int64_t ts,
+                                     const char *event_class_name, const char *hostname,
+                                     int64_t vpid, uint64_t vtid) {
+  hpt_t hpt = {hostname, vpid, vtid};
+  try {
+    send_host_message(btx_handle, usr_data, ts, event_class_name,
+                      hostname, vpid, vtid, false);
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+      "entry ts not found for thread in exit cudaError_absent callback");
+  }
+}
```

Comment:
Cannot append, so can remove some code :)

Reply:
Discussed this a lot on slack, decided to drop verbose try/catch error handling in favor of THAPI_AT macro along with valgrind to get stack traces.